File: Basic/Python/1005.py
'''
@Author: kurumi
@Date: 2025-08-23 13:48
@Description: 1005 继续(3n+1)猜想
'''
from typing import List, Set

calNums: Set[int] = set()
keyNums: List[bool] = [False] * 200

def isOdd(n: int) -> bool:
    return n % 2 != 0

def isEven(n: int) -> bool:
    return n % 2 == 0

def Callatz(n: int):
    while n != 1:
        if n in calNums:
            keyNums[n] = False
            return
        calNums.add(n)
        if isOdd(n):
            n = (3 * n + 1) // 2
        else:
            n = n // 2

def main():
    n: int = int(input())
    res: List[int] = []
    nums: List[int] = list(map(int, input().split()))
    for num in nums:
        keyNums[num] = True
    for num in nums:
        Callatz(num)
    for num in nums:
        if keyNums[num]:
            res.append(num)    
    res.sort(reverse=True)
    print(" ".join(map(str, res)))    

if __name__ == "__main__":
    main()

File: Basic/JS/1005.js
/**
 * @author: kurumi
 * @date: 2025-08-24 13:50
 * @description: 1005 继续(3n+1)猜想
 */
const rl = require("readline").createInterface({input: process.stdin});
const iter = rl[Symbol.asyncIterator]();
const readline = async () => (await iter.next()).value;

const calNums = new Set();
const keyNums = new Array(200).fill(false);

const isOdd = n => n % 2 != 0;

const isEven = n => n % 2 == 0;

const Callatz = n => {
  while (n != 1) {
    if (calNums.has(n)) {
      keyNums[n] = false;
      return;
    }
    calNums.add(n);
    if (isOdd(n)) {
      n = Math.ceil((3 * n + 1) / 2);
    } else {
      n /= 2;
    }
  }
}

void (async function (){
  const n = parseInt(await readline());
  res = new Array();
  nums = (await readline()).split(" ").map(s => parseInt(s));
  for (let num of nums) {
    keyNums[num] = true;
  }
  nums.forEach(num => Callatz(num));
  nums.forEach(num => {
    if (keyNums[num]) res.push(num);
  });
  res.sort((a, b) => b - a);
  console.log(res.join(" "));
})();